observatory-dags/observatory/dags/telescopes/crossref_events.py

```diff
@@ -0,0 +1,388 @@
+        """ Task to download the CrossrefEventsRelease release.
+
+        :param release: a CrossrefEventsRelease instance.
+        :return: None.
+        """
+        # Download release
+        release.download()
+
+    def upload_downloaded(self, release: CrossrefEventsRelease, **kwargs):
+        """ Task to upload the downloadeded CrossrefEventsRelease release.
+
+        :param release: a CrossrefEventsRelease instance.
+        :return: None.
+        """
+        # Upload each downloaded release
+        upload_files_from_list(release.download_files, release.download_bucket)
+
+    def transform(self, release: CrossrefEventsRelease, **kwargs):
+        """ Task to transform the CrossrefEventsRelease release.
+
+        :param release: an CrossrefEventsRelease instance.
+        :return: None.
+        """
+        # Transform each release
+        release.transform()
+
+
+def extract_events(url: str, events_path: str, next_cursor: str = None, success: bool = True) -> \
+        Tuple[bool, Union[str, None], Union[int, None]]:
+    """
+    Extract the events from the given url until no new cursor is returned or a RetryError occurs. The extracted events
+    are appended to a json file, with 1 list per request.
+
+    :param url: The crossref events api url
+    :param events_path: Path to the file in which events are stored
+    :param next_cursor: The next cursor, this is in the response of the api
+    :param success: Whether all events were extracted successfully or an error occurred
+    :return: success, next_cursor and number of total events
+    """
+    headers = {
+        'User-Agent': 'Observatory Platform (+https://github.com/The-Academic-Observatory/observatory-platform; '
+                      f'mailto:{CrossrefEventsTelescope.MAILTO})'
+    }
+    tmp_url = url + f'&cursor={next_cursor}' if next_cursor else url
+    try:
+        response = retry_session(num_retries=15, backoff_factor=0.5, status_forcelist=(500, 400)).get(tmp_url,
+                                                                                                      headers=headers)
+    except RetryError:
+        return False, next_cursor, None
+    if response.status_code == 200:
+        response_json = response.json()
+        total_events = response_json['message']['total-results']
+        events = response_json['message']['events']
+        next_cursor = response_json['message']['next-cursor']
+        # write events so far
+        with open(events_path, 'a') as f:
+            f.write(json.dumps(events) + '\n')
+        if next_cursor:
+            success, next_cursor, total_events = extract_events(url, events_path, next_cursor, success)
+            return success, next_cursor, total_events
+        else:
+            return True, None, total_events
+    else:
+        raise ConnectionError(f"Error requesting url: {tmp_url}, response: {response.text}")
+
+
+def download_events_batch(release: CrossrefEventsRelease, i: int) -> list:
```

Comment:
This method might as well be part of the release object, since release is passed in as the first parameter.